pkg/stream/edgedmetricsconnection_test.go

```diff
@@ -92,3 +158,141 @@ func TestMetricsConnection_CloseReadChannel(t *testing.T) {
+	}
+
+	metricsConn.Stop <- struct{}{}
+	metricsConn.Stop <- struct{}{}
+
+	metricsConn.CleanChannel()
+
+	assert.Equal(0, len(metricsConn.Stop))
+}
+
+func TestMetricsConnection_receiveFromCloudStream(t *testing.T) {
+	assert := assert.New(t)
+
+	stop := make(chan struct{}, 1)
+
+	metricsConn := &EdgedMetricsConnection{
+		MessID:   uint64(100),
+		ReadChan: make(chan *Message, 3),
+	}
+
+	removeConnMsg := NewMessage(metricsConn.MessID, MessageTypeRemoveConnect, nil)
+
+	metricsConn.ReadChan <- removeConnMsg
+
+	close(metricsConn.ReadChan)
+
+	metricsConn.receiveFromCloudStream(stop)
+
+	assert.Equal(1, len(stop))
+}
+
+func TestMetricsConnection_write2CloudStream(t *testing.T) {
+	assert := assert.New(t)
+
+	mockTunneler := &MockStreamTunneler{}
+	stop := make(chan struct{}, 1)
+
+	responseBody := NewMockResponseBody("line1\nline2\nline3")
+	mockResponse := &http.Response{
+		Body: responseBody,
+	}
+
+	metricsConn := &EdgedMetricsConnection{
+		MessID: uint64(100),
+	}
+
+	go metricsConn.write2CloudStream(mockTunneler, mockResponse, stop)
+
+	time.Sleep(100 * time.Millisecond)
+
+	assert.Equal(3, len(mockTunneler.Messages))
+	assert.Equal(MessageTypeData, mockTunneler.Messages[0].MessageType)
+	assert.Contains(string(mockTunneler.Messages[0].Data), "line1")
+
+	assert.Equal(1, len(stop))
+}
+
+func TestMetricsConnection_write2CloudStream_WriteError(t *testing.T) {
+	assert := assert.New(t)
+
+	mockTunneler := &MockStreamTunneler{
+		WriteErr: errors.New("tunnel write error"),
+	}
+	stop := make(chan struct{}, 1)
+
+	responseBody := NewMockResponseBody("test data for tunnel")
+	mockResponse := &http.Response{
+		Body: responseBody,
+	}
+
+	metricsConn := &EdgedMetricsConnection{
+		MessID: uint64(100),
+	}
+
+	go metricsConn.write2CloudStream(mockTunneler, mockResponse, stop)
+
+	time.Sleep(100 * time.Millisecond)
+
+	assert.Equal(1, len(stop))
+}
+
+func TestMetricsConnection_Serve(t *testing.T) {
+	assert := assert.New(t)
+
+	metricsConn := &EdgedMetricsConnection{
+		MessID:   uint64(100),
+		ReadChan: make(chan *Message, 10),
+		Stop:     make(chan struct{}, 1),
+		URL:      url.URL{Scheme: "https", Host: "example.com", Path: "/metrics"},
+		Header:   http.Header{},
+	}
+
+	mockTunneler := &MockStreamTunneler{}
+
+	responseBody := NewMockResponseBody("test metric data")
+	mockResponse := &http.Response{
+		StatusCode: 200,
+		Body:       responseBody,
+	}
+
+	patchNewRequest := gomonkey.ApplyFunc(http.NewRequest,
+		func(method, url string, body io.Reader) (*http.Request, error) {
+			return &http.Request{
+				Header: http.Header{},
+			}, nil
+		})
+	defer patchNewRequest.Reset()
+
+	patchDo := gomonkey.ApplyMethod(reflect.TypeOf(&http.Client{}), "Do",
+		func(_ *http.Client, _ *http.Request) (*http.Response, error) {
+			return mockResponse, nil
+		})
+	defer patchDo.Reset()
+
+	go func() {
+		time.Sleep(100 * time.Millisecond)
+		metricsConn.Stop <- struct{}{}
+	}()
+
+	err := metricsConn.Serve(mockTunneler)
+
+	assert.NoError(err)
```

Comment:
Next time, you can use require(`github.com/stretchr/testify/require`) and assert flexibly. Functions in the require package will break the test, but functions in the assert package will not.